src/assisted_installer_controller/assisted_installer_controller.go

```diff
@@ -238,48 +238,91 @@ func (c controller) UpdateBMHs(wg *sync.WaitGroup) {
 			continue
 		}
 
-		allUpdated := c.updateBMHStatus(bmhs)
+		machines, err := c.unallocatedMachines(bmhs)
+		if err != nil {
+			c.log.WithError(err).Errorf("Failed to get machines")
+			continue
+		}
+
+		allUpdated := c.updateBMHs(bmhs, machines)
 		if allUpdated {
 			c.log.Infof("Updated all the BMH CRs, finished successfully")
 			return
 		}
 	}
 }
 
-func (c controller) updateBMHStatus(bmhList metal3v1alpha1.BareMetalHostList) bool {
+func (c controller) unallocatedMachines(bmhList metal3v1alpha1.BareMetalHostList) (*mapiv1beta1.MachineList, error) {
+	machineList, err := c.kc.ListMachines()
+	if err != nil {
+		return nil, err
+	}
+	// filter out machines that are not an allocatable worker
+	for m := 0; m < len(machineList.Items); m++ {
+		role, ok := machineList.Items[m].Labels["machine.openshift.io/cluster-api-machine-role"]
+		remove := (!ok || role != "worker")
+		for _, bmh := range bmhList.Items {
+			if bmh.Spec.ConsumerRef != nil && bmh.Spec.ConsumerRef.Name == machineList.Items[m].Name {
+				remove = true
+			}
+		}
+		if remove {
+			machineList.Items = append(machineList.Items[:m], machineList.Items[m+1:]...)
+			m-- // -1 as the slice just got shorter
+		}
+	}
+	return machineList, nil
+}
+
+func (c controller) updateBMHs(bmhList metal3v1alpha1.BareMetalHostList, machineList *mapiv1beta1.MachineList) bool {
 	allUpdated := true
+
 	for i := range bmhList.Items {
 		bmh := bmhList.Items[i]
 		c.log.Infof("Checking bmh %s", bmh.Name)
 		annotations := bmh.GetAnnotations()
 		content := []byte(annotations[metal3v1alpha1.StatusAnnotation])
-		if annotations[metal3v1alpha1.StatusAnnotation] == "" {
+
+		if annotations[metal3v1alpha1.StatusAnnotation] == "" && bmh.Spec.ConsumerRef != nil {
```

Comment:
Seems that the only way for this method to return true (allUpdated) is if all bmh will get true in this if.
Does that make sense?

Reply:
Yes, I believe it is correct.
It scans through all bmh and makes sure that 
1. the status annotation has been moved
2. the consumerRef has been set

"1." is as before and "2." will already be true for masters, and we are setting for workers.
The function gets continuously called so even if we set them all in one run, it will be called once more for a check and then it will return true.